File: sdk/ts/grpc-test.ts
// TS SDK gRPC 契约测试：GetConfig / GetItem / Watch / ListMembers（走 :8383）。
import { ConfigClient } from './src/index.ts';

const GRPC = process.env.DSH_GRPC || '127.0.0.1:8383';
const HTTP = process.env.DSH_HTTP || 'http://127.0.0.1:8384';
const P = process.env.DSH_PROJECT || 'sdk-project';

async function main() {
  // gRPC 优先端点（design §3.1 Endpoint{grpc?,http?}）
  const c = new ConfigClient([{ grpc: GRPC, http: HTTP }], { token: process.env.DSH_TOKEN });

  const snap = await c.get(P, 'dev');
  const host = (snap.groups as any)?.redis?.host;
  console.log('[ts-grpc] get ok: version=' + snap.version + ' host=' + host);
  if (!host || typeof host !== 'string') throw new Error('get value mismatch');

  const item = await c.getItem(P, 'dev', 'redis', 'host');
  console.log('[ts-grpc] getItem ok: ' + item);
  if (item !== host) throw new Error('getItem mismatch');

  // D-TEST：ListMembers 真断言——dev-single 下应为 FailedPrecondition（gRPC code 9）
  try {
    const members = await c.listMembers();
    if (members.length) throw new Error('dev-single 不应有成员，got ' + members.length);
    console.log('[ts-grpc] listMembers dev-single 返回空列表（契约语义：非集群可用）');
  } catch (e: any) {
    if (e && (e.code === 'GRPC_9' || e.code === 'GRPC_FAILED_PRECONDITION')) {
      console.log('[ts-grpc] listMembers dev-single → FailedPrecondition ✅');
    } else {
      throw e;
    }
  }

  await new Promise<void>((resolve, reject) => {
    const ctrl = new AbortController();
    const timer = setTimeout(() => {
      ctrl.abort();
      reject(new Error('watch timeout'));
    }, 10000);
    c.watch(P, 'dev', (e) => {
      if (e.version > snap.version) {
        clearTimeout(timer);
        ctrl.abort();
        c.close();
        console.log('[ts-grpc] watch event: v' + e.version + ' ' + e.ty);
        resolve();
      }
    }, ctrl.signal);
  });
  console.log('[ts-grpc] PASS');
}

main().catch((e) => {
  console.error('[ts-grpc] FAIL:', e);
  process.exit(1);
});

File: sdk/python/grpc-test.py
"""Python SDK gRPC 契约测试：get / get_item / watch / list_members（:8383）。"""

import os
import sys
import threading
import time

import grpc

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from config_client import ConfigClient

GRPC = os.environ.get("DSH_GRPC", "127.0.0.1:8383")
HTTP = os.environ.get("DSH_HTTP", "http://127.0.0.1:8384")
PROJECT = os.environ.get("DSH_PROJECT", "sdk-project")


def main():
    c = ConfigClient([{"grpc": GRPC, "http": HTTP}])
    snap = c.get(PROJECT, "dev")
    host = snap.get("groups", {}).get("redis", {}).get("host")
    print("[py-grpc] get ok: version=%s host=%s" % (snap["version"], host))
    if not host:
        raise SystemExit("[py-grpc] FAIL value mismatch")

    item = c.get_item(PROJECT, "dev", "redis", "host")
    print("[py-grpc] get_item ok: %s" % item)
    if item != host:
        raise SystemExit("[py-grpc] FAIL get_item mismatch")

    # D-TEST：ListMembers 真断言——dev-single 下应为 FailedPrecondition（gRPC code 9）
    try:
        members = c.list_members()
        if members:
            raise SystemExit("[py-grpc] FAIL list_members 不应在 dev-single 返回成员: %r" % members)
        print("[py-grpc] list_members dev-single 返回空列表（契约语义：非集群可用）")
    except Exception as e:
        # grpc.RpcError：code() 方法返回 StatusCode 枚举（跨版本 .value 形状不一，直接枚举比较）
        code_fn = getattr(e, "code", None)
        if callable(code_fn) and code_fn() == grpc.StatusCode.FAILED_PRECONDITION:
            print("[py-grpc] list_members dev-single → FailedPrecondition ✅")
        else:
            raise SystemExit("[py-grpc] FAIL list_members unexpected: %r" % e)

    stop = threading.Event()
    got = []

    def on_event(e):
        if e["version"] > snap["version"]:
            got.append(e["version"])

    t = threading.Thread(target=lambda: c.watch(PROJECT, "dev", on_event, stop), daemon=True)
    t.start()
    deadline = time.time() + 15
    while time.time() < deadline:
        if got:
            print("[py-grpc] watch event: v%d" % got[0])
            stop.set()
            print("[py-grpc] PASS")
            return
        time.sleep(0.1)
    stop.set()
    raise SystemExit("[py-grpc] FAIL watch timeout")


if __name__ == "__main__":
    main()
